### ISSUE TITLE
Improvement ideas

### ISSUE THREAD

**Issue by balthazarneveu** [balthazarneveu/molecule-retrieval-using-nlp#2]
# TODOS to improve
- [ ] Memory profiling
- [ ] Torch compile
- [x] Pretrain the LLM, freeze it then focus on molecule graphs and get a bigger GCN = use large graphs batch size
- [x] Ensembling :free:
- [ ] Alternate GCN and LLM gradient descent / stop gradients
- [ ] Gradient accumulations
- [ ] Augment text and molecules?
- [ ] Tiplet Loss, Contrastive loss, [NT-Xent?](https://www.youtube.com/watch?v=u-X_nZRsn5M)

![image](https://github.com/balthazarneveu/molecule-retrieval-using-nlp/assets/41070742/c458c489-498f-43ad-a0de-fa03e7ed0bb3)

# Potential easy wins
#8 
- [x] Freeze language model weights (maybe we store all BERT's weights and retrain it, which is absurd)
- [x] Store BERT embeddings of the whole text training set and feed it to an adapter network to output more relevant embeddings
- [ ] Handcrafted features: Add the number of links and histogram of atoms?

# Deeper and bigger graph encoder
#13 
- Play with the GNN architecture. 

# Auto-encoder for molecules
- AE for graphs, then extract latent, feed it into a NN to optimize the contrastive loss
